src/indexable.cr

```diff
@@ -0,0 +1,469 @@
+  # ```
+  def equals?(other)
+    return false if size != other.size
+    each_with_index do |item, i|
+      return false unless yield(item, other[i])
+    end
+    true
+  end
+
+  # Returns the first element of `self` if it's not empty, or raises `IndexError`.
+  #
+  # ```
+  # ([1, 2, 3]).first   # => 1
+  # ([] of Int32).first # => raises IndexError
+  # ```
+  def first
+    first { raise IndexError.new }
+  end
+
+  # Returns the first element of `self` if it's not empty, or the given block's value.
+  #
+  # ```
+  # ([1, 2, 3]).first { 4 }   # => 1
+  # ([] of Int32).first { 4 } # => 4
+  # ```
+  def first
+    size == 0 ? yield : unsafe_at(0)
+  end
+
+  # Returns the first element of `self` if it's not empty, or `nil`.
+  #
+  # ```
+  # ([1, 2, 3]).first?   # => 1
+  # ([] of Int32).first? # => nil
+  # ```
+  def first?
+    first { nil }
+  end
+
+  # Returns a hash code based on `self`'s size and elements.
+  #
+  # See `Object#hash`.
+  def hash
+    reduce(31 * size) do |memo, elem|
+      31 * memo + elem.hash
+    end
+  end
+
+  # Returns the last element of `self` if it's not empty, or raises `IndexError`.
+  #
+  # ```
+  # ([1, 2, 3]).last   # => 3
+  # ([] of Int32).last # => raises IndexError
+  # ```
+  def last
+    last { raise IndexError.new }
+  end
+
+  # Returns the last element of `self` if it's not empty, or the given block's value.
+  #
+  # ```
+  # ([1, 2, 3]).last { 4 }   # => 3
+  # ([] of Int32).last { 4 } # => 4
+  # ```
+  def last
+    size == 0 ? yield : unsafe_at(size - 1)
+  end
+
+  # Returns the last element of `self` if it's not empty, or `nil`.
+  #
+  # ```
+  # ([1, 2, 3]).last?   # => 1
+  # ([] of Int32).last? # => nil
+  # ```
+  def last?
+    last { nil }
+  end
+
+  # Same as `#each`, but works in reverse.
+  def reverse_each(&block)
+    (size - 1).downto(0) do |i|
+      yield unsafe_at(i)
+    end
+    self
+  end
+
+  # Returns an `Iterator` over the elements of `self` in reverse order.
+  def reverse_each
+    ReverseItemIterator(self, T).new(self)
+  end
+
+  def rindex(value)
+    rindex { |elem| elem == value }
+  end
+
+  def rindex
+    (size - 1).downto(0) do |i|
+      if yield unsafe_at(i)
+        return i
+      end
+    end
+    nil
+  end
+
+  # Returns a random element from `self`, using the given *random* number generator.
+  # Raises IndexError if `self` is empty.
+  #
+  # ```
+  # a = [1, 2, 3]
+  # a.sample                # => 2
+  # a.sample                # => 1
+  # a.sample(Random.new(1)) # => 3
+  # ```
+  def sample(random = Random::DEFAULT)
+    raise IndexError.new if size == 0
+    unsafe_at(random.rand(size))
+  end
+
+  # Returns *n* number of random elements from `self`, using the given *random* number generator.
+  # Raises IndexError if `self` is empty.
+  #
+  # ```
+  # a = [1, 2, 3]
+  # a.sample(2)                # => [2, 1]
+  # a.sample(2, Random.new(1)) # => [1, 3]
+  # ```
+  def sample(n : Int, random = Random::DEFAULT)
```

Comment:
@bcardiff I thought about that too, it's the only method that returns an Array. We can probably keep it in Array for now.